lib/css-converter.js

```diff
@@ -0,0 +1,257 @@
+ *
+ * @param {*} str
+ */
+function escapeRegexLiterals (str) {
+  // The no-useless-escape regex rule is wrong when it's in a Regex. These escapes are intentional.
+  // eslint-disable-next-line no-useless-escape
+  return str.replace(/[\[\^\$\.\|\?\*\+\(\)\\]/g, (tok) => `\\${tok}`);
+}
+
+/**
+ * Add android:id/ to beginning of string if it's not there already
+ * @param {*} str
+ */
+function prependAndroidId (str) {
+  if (!str.startsWith('android:id/')) {
+    return `android:id/${str}`;
+  }
+  return str;
+}
+
+/**
+ * Convert a CSS attribute into a UiSelector method call
+ * @param {*} cssAttr CSS attribute object
+ */
+function parseAttr (cssAttr) {
+  if (cssAttr.valueType && cssAttr.valueType !== 'string') {
+    log.errorAndThrow(`Could not parse '${cssAttr.name}=${cssAttr.value}'.` +
+      `Unsupported attribute type '${cssAttr.valueType}'`);
+  }
+  const attrName = assertGetAttrName(cssAttr);
+  const methodName = toSnakeCase(attrName);
+  if (booleanAttrs.includes(attrName)) {
+    return `.${methodName}(${assertGetBool(cssAttr)})`;
+  }
+
+  if (strAttrs.includes(attrName)) {
+    let value = cssAttr.value || '';
+    if (attrName === 'resource-id') {
+      value = prependAndroidId(value);
+    }
+    if (value === '') {
+      return `.${methodName}Matches("")`;
+    } else {
+      switch (cssAttr.operator) {
+        case '=':
+          return `.${methodName}("${value}")`;
+        case '*=':
+          if (['description', 'text'].includes(attrName)) {
+            return `.${methodName}Contains("${value}")`;
+          }
+          return `.${methodName}Matches("${escapeRegexLiterals(value)}")`;
+        case '^=':
+          if (['description', 'text'].includes(attrName)) {
+            return `.${methodName}StartsWith("${value}")`;
+          }
+          return `.${methodName}Matches("^${escapeRegexLiterals(value)}")`;
+        case '$=':
+          return `.${methodName}Matches("${escapeRegexLiterals(value)}$")`;
+        case '~=':
+          return `.${methodName}Matches("${getWordMatcherRegex(value)}")`;
+        default:
+          log.errorAndThrow(`Unsupported CSS attribute operator '${cssAttr.operator}'`);
+      }
+    }
+  }
+}
+
+/**
+ * Convert a CSS pseudo class to a UiSelector
+ * @param {*} cssPseudo CSS Pseudo class
+ */
+function parsePseudo (cssPseudo) {
+  if (cssPseudo.valueType && cssPseudo.valueType !== 'string') {
+    log.errorAndThrow(`Could not parse '${cssPseudo.name}=${cssPseudo.value}'. ` +
+      `Unsupported css pseudo class value type: '${cssPseudo.valueType}'`);
+  }
+
+  const pseudoName = assertGetAttrName(cssPseudo);
+
+  if (booleanAttrs.includes(pseudoName)) {
+    return `.${toSnakeCase(pseudoName)}(${assertGetBool(cssPseudo)})`;
+  }
+
+  if (numericAttrs.includes(pseudoName)) {
+    return `.${pseudoName}(${cssPseudo.value})`;
+  }
+}
+
+/**
+ * Convert a CSS rule to a UiSelector
+ * @param {*} cssRule CSS rule definition
+ */
+function parseCssRule (cssRule) {
```

Comment:
I think it makes sense to have a class for such transformation rather than a set of functions.

Reply:
@mykola-mokhnach I prefer the procedural approach myself, but that's a much larger debate.

Reply:
having a class could help to avoid passing the css object (and possibly many other shared args) to all the functions

Reply:
also, this could be refactored to have a main class that performs css expression validation and tokenization and then passing a tranformer class/function to it. This class/function could be responsible for the expression tranformation itself to. let say UiSelector or class chain expression. This way we could extract the common part of the code and possibly reuse it by moving to appium-support module. So, only the driver-specific transformation method would stay in the driver source.

Reply:
I chose this approach because of my personal preference ... I find the testing simpler and I personally lean towards avoiding the cognitive overhead of thinking about "state". That is it to say, when I'm writing and troubleshooting functions, I prefer the functions to be "pure" so that I only have to think about the inputs (the args) and the output and that I don't have to think about the state of the object under test.

I still believe there are many cases where using classes/objects is the right approach. I won't be advocating that we port Appium to Haskell any time soon. I like that Javascript affords the choice between functional programming and object-oriented :)

Reply:
It is fine to use Haskell-based approach. For me it's more important to have it properly designed for architectural perspective. E.g. that we could easily separate different entities (tokenization, validation and transformation) and then implement the feature in different drivers without copying the same parts of code here and there

Reply:
For this feature, my plan is to not share any code between the drivers and just copy-paste + refactor.

Normally this is terrible coding practice, but I'm making an exception because this is based on the CSS selector standard, which isn't going to be introducing breaking changes any time soon and it's very unlikely that we'll ever need to make alterations to the code that would have to be ported to all drivers. I would rather each driver do it's own implementation.